File: 1306 Jump Game III/1306.js
/**
 * @param {number[]} arr
 * @param {number} start
 * @return {boolean}
 */
var canReach = function(arr, start) {
    if (arr.length === 1) {
        return true;
    }
    return dfs(arr, start);
};

var dfs = function(arr, index) {
    if (index < 0 || index >= arr.length || arr[index] >= arr.length) {
        return false;
    }
    if (arr[index] === 0) {
        return true;
    }
    let jump = arr[index]
    arr[index] += arr.length
    return dfs(arr, index - jump) || dfs(arr, index + jump);
} 

File: 1306 Jump Game III/1306.py
class Solution:
    def canReach(self, arr: List[int], start: int) -> bool:
        if len(arr) == 1:
            return True
        return dfs(arr, start)
    
def dfs(arr, index):
    if index < 0 or index >= len(arr) or arr[index] >= len(arr):
        return False
    if arr[index] == 0:
        return True
    jump = arr[index]
    arr[index] += len(arr)
    return dfs(arr, index - jump) or dfs(arr, index + jump)